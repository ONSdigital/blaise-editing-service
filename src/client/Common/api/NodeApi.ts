import axios from 'axios';
import notFound from '../../../server/helpers/axiosHelper';
import { Survey } from '../../../common/interfaces/surveyInterface';

async function getDataFromNode<T>(url: string, notFoundError: string): Promise<T> {
  try {
    const response = await axios.get(url);

    return response.data;
  } catch (error) {
    if (notFound(error)) {
      throw new Error(notFoundError);
    }
    throw new Error('Unable to complete request, please try again in a few minutes');
  }
}

export default async function getSurveys(): Promise<Survey[]> {
  return getDataFromNode('/api/surveys', 'Unable to find surveys, please contact Richmond Rice');
}
