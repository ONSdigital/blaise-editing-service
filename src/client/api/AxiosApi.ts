import axios, { AxiosRequestConfig } from 'axios';
import { AuthManager } from 'blaise-login-react-client';
import notFound from '../../server/helpers/axiosHelper';

function axiosConfig(): AxiosRequestConfig {
  const authManager = new AuthManager();
  return {
    headers: {
      'Content-Type': 'application/json',
      ...authManager.authHeader(),
    },
  };
}

export async function getDataFromNode<T>(url: string, notFoundError: string): Promise<T> {
  try {
    const response = await axios.get(url, axiosConfig());

    return response.data;
  } catch (error) {
    if (notFound(error)) {
      throw new Error(notFoundError);
    }
    throw new Error('Unable to complete request, please try again in a few minutes');
  }
}

export async function patchDataToNode(url: string, payload: any, notFoundError: string): Promise<void> {
  try {
    await axios.patch(url, payload, axiosConfig());
  } catch (error) {
    if (notFound(error)) {
      throw new Error(notFoundError);
    }
    throw new Error('Unable to complete request, please try again in a few minutes');
  }
}
