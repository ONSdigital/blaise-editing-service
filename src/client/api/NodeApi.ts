import axios from 'axios';

import { CaseEditInformation } from 'blaise-api-node-client/lib/cjs/interfaces/case';
import { User } from 'blaise-api-node-client/lib/cjs/interfaces/user';
import notFound from '../../server/helpers/axiosHelper';
import { Survey } from '../../common/interfaces/surveyInterface';
import { SupervisorInformation } from '../Interfaces/supervisorInterface';
import { EditorInformation } from '../Interfaces/editorInterface';
import mapEditorInformation from '../Mappers/editorInformaitionMapper';
import mapSupervisorInformation from '../Mappers/supervisorInformationMapper';
import { CaseSummaryDetails } from '../../common/interfaces/caseInterface';
import { CasesNotAllocatedInformation } from '../Interfaces/caseAllocationInterface';
import mapCasesNotAllocated from '../Mappers/caseAllocationMapper';
// import { caseSummaryDetailsMockObject } from '../../test/server/mockObjects/CaseMockObject';

async function getDataFromNode<T>(url: string, notFoundError: string): Promise<T> {
  try {
    const response = await axios.get(url);

    return response.data;
  } catch (error) {
    if (notFound(error)) {
      throw new Error(notFoundError);
    }
    throw new Error('Unable to complete request, please try again in a few minutes');
  }
}

export async function getSurveys(userRole: string): Promise<Survey[]> {
  return getDataFromNode(`/api/surveys?userRole=${userRole}`, 'Unable to find surveys, please contact Richmond Rice');
}

export async function getCaseSummary(questionnaireName: string, caseId: string): Promise<CaseSummaryDetails> {
  return getDataFromNode(`/api/questionnaires/${questionnaireName.toUpperCase()}/cases/${caseId}/summary`, 'The questionnaire is no longer available');
}

async function getCaseEditInformation(questionnaireName: string, userRole: string) {
  // TODO Fix the URL upper
  return getDataFromNode<CaseEditInformation[]>(`/api/questionnaires/${questionnaireName.toUpperCase()}/cases/edit?userRole=${userRole}`, 'Unable to find case edit information, please contact Richmond Rice');
}

export async function getEditorInformation(questionnaireName: string, editorUsername:string, editorRole: string): Promise<EditorInformation> {
  const caseEditInformationList = await getCaseEditInformation(questionnaireName, editorRole);
  const caseEditInformationListForEditor = caseEditInformationList.filter((caseEditInformation) => caseEditInformation.assignedTo === editorUsername);

  return mapEditorInformation(caseEditInformationListForEditor);
}

export async function getSupervisorEditorInformation(questionnaireName: string, supervisorRole: string, editorRole: string): Promise<SupervisorInformation> {
  const caseEditInformationList = await getCaseEditInformation(questionnaireName, supervisorRole);
  const editors = await getDataFromNode<User[]>(`/api/users?userRole=${editorRole}`, 'Unable to find user information, please contact Richmond Rice');

  return mapSupervisorInformation(caseEditInformationList, editors);
}

export async function getCasesNotAllocatedInformation(questionnaireName: string, supervisorRole: string, editorRole: string): Promise<CasesNotAllocatedInformation> {
  const caseEditInformationList = await getCaseEditInformation(questionnaireName, supervisorRole);
  const editors = await getDataFromNode<User[]>(`/api/users?userRole=${editorRole}`, 'Unable to find user information, please contact Richmond Rice');

  return mapCasesNotAllocated(caseEditInformationList, editors);
}
