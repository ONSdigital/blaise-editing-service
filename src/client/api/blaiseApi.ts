import axios from 'axios';
import { IQuestionnaire } from 'blaise-api-node-client';
import { ICaseDetails } from '../../server/interfaces/case.details.interface';

export async function getQuestionnaires(): Promise<IQuestionnaire[]> {
  const response = await axios.get('/api/questionnaires');
  // if 404 throw error with not found message

  // if 500 throw error with default message

  return response.data;
}

export async function getListOfCases(questionnaireName: string): Promise<ICaseDetails[]> {
  const response = await axios.get(`/api/questionnaires/${questionnaireName}/cases`);
  return response.data;
}
