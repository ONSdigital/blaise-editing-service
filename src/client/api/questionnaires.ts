import axios from 'axios';
import { GetQuestionnaires } from '../../shared/responses/GetQuestionnaires';

export default async function getQuestionnaires(): Promise<GetQuestionnaires> {
  const response = await axios.get('/api/questionnaires');
  return response.data;
}
