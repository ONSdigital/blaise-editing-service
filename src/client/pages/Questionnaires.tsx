import axios from 'axios';
import { useEffect, useState } from 'react';

export default function Questionnaires() {
  const [questionnaires, setQuestionnaires] = useState<{ name:string }[]>([]);

  useEffect(() => {
    async function fetchQuestionnaires() {
      const response = await axios.get('http://localhost:3100/questionnaires');
      return response.data;
    }

    fetchQuestionnaires().then(setQuestionnaires);
  }, []);
  return (
    <div>
      {questionnaires.map((questionnaire) => {
        if ('name' in questionnaire) return questionnaire.name;
        return null;
      })}
    </div>
  );
}
