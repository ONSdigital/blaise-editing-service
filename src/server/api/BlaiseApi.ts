import BlaiseClient, {
  CaseResponse, CaseStatus, Questionnaire, QuestionnaireReport,
} from 'blaise-api-node-client';
import { Configuration } from '../interfaces/configurationInterface';
import { QuestionnaireAllocation } from '../../common/interfaces/surveyInterface';
import mapQuestionnaireAllocation from '../mappers/questionnaireMapper';

export default class BlaiseApi {
  config: Configuration;

  blaiseApiClient: BlaiseClient;

  constructor(config: Configuration, blaiseApiClient: BlaiseClient) {
    this.config = config;
    this.blaiseApiClient = blaiseApiClient;
    this.getCaseStatus = this.getCaseStatus.bind(this);
    this.getCaseStatus = this.getCaseStatus.bind(this);
    this.getQuestionnaires = this.getQuestionnaires.bind(this);
    this.getReportsForQuestionnaires = this.getReportsForQuestionnaires.bind(this);
  }

  async getCaseStatus(questionnaireName: string): Promise<CaseStatus[]> {
    return this.blaiseApiClient.getCaseStatus(this.config.ServerPark, questionnaireName);
  }

  async getCase(questionnaireName: string, caseId: string): Promise<CaseResponse> {
    return this.blaiseApiClient.getCase(this.config.ServerPark, questionnaireName, caseId);
  }

  async getQuestionnaires(): Promise<QuestionnaireAllocation[]> {
    const questionnaires = await this.blaiseApiClient.getQuestionnaires(this.config.ServerPark) as QuestionnaireAllocation[];
    const questionnairesReportData = await this.getReportsForQuestionnaires(questionnaires);

    return mapQuestionnaireAllocation(questionnaires, questionnairesReportData);
  }

  private async getReportsForQuestionnaires(questionnaires: Questionnaire[]): Promise<QuestionnaireReport[]> {
    const allocationFieldIds = ['allocation.toeditor'];
    const reports: Promise<QuestionnaireReport>[] = [];

    questionnaires.forEach((questionnaire) => {
      reports.push(this.blaiseApiClient.getQuestionnaireReportData(this.config.ServerPark, questionnaire.name, allocationFieldIds));
    });

    return Promise.all(reports);
  }
}
