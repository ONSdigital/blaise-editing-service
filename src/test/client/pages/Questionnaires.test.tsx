import { render, waitForElementToBeRemoved, screen } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';
import { IQuestionnaire, QuestionnaireListMockObject } from 'blaise-api-node-client';
import Questionnaires from '../../../client/pages/Questionnaires';
import { getQuestionnaires } from '../../../client/api/blaiseApi';

jest.mock('../../../client/api/blaiseApi');

const getQuestionnairesMock = getQuestionnaires as jest.Mock<Promise<IQuestionnaire[]>>;

describe('Given there are questionnaires available in blaise', () => {
  beforeEach(() => {
    getQuestionnairesMock.mockImplementation(() => Promise.resolve(QuestionnaireListMockObject));
  });

  afterEach(() => {
    getQuestionnairesMock.mockReset();
  });

  it('should render the page correctly', async () => {
    // act
    const wrapper = render(
      <BrowserRouter>
        <Questionnaires />
      </BrowserRouter>,
    );

    await waitForElementToBeRemoved(screen.getByText('Loading...'));

    // assert
    expect(wrapper).toMatchSnapshot();
  });

  it('should display a list of the expected questionnaires', async () => {
    // act
    const { queryByText } = render(
      <BrowserRouter>
        <Questionnaires />
      </BrowserRouter>,
    );

    await waitForElementToBeRemoved(screen.getByText('Loading...'));

    // assert
    QuestionnaireListMockObject.forEach((questionnaire) => {
      expect(queryByText(questionnaire.name)).toBeInTheDocument();
    });
  });

  describe('Given there are no questionnaires available in blaise', () => {
    it('should display a message telling the user there are no questionnaires', async () => {
      // arrange
      // set mock to return an empty array

      // act
      // render the page

      // assert
      // check the message is on the screen
      // check the message is on the screen and that it is in an information panel maybe? Snapshot test as well?
    });
  });

  // this one test should be enough cover all error scenarios - the granular errors can be tested against the blaise api file
  describe('Given there the blaise rest api is not available', () => {
    it('should display an error message telling the user to try again in a few minutes', async () => {
      // arrange
      // set mock to throw an error with the message

      // act
      // render the page

      // assert
      // check the message is on the screen and that it is in an error panel maybe? Snapshot test as well?
    });
  });
});
