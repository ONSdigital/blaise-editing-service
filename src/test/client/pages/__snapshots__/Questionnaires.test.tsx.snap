// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders questionnaires page questionnaires list page matches Snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table
        class="ons-table"
        data-testid="questionnaire-table"
        id="questionnaire-table"
      >
        <thead
          class="ons-table__head"
        >
          <tr
            class="ons-table__row"
          >
            <th
              class="ons-table__header"
              scope="col"
            >
              Questionnaire name
            </th>
            <th
              class="ons-table__header"
              scope="col"
            >
              Server park
            </th>
            <th
              class="ons-table__header"
              scope="col"
            >
              Number of cases
            </th>
          </tr>
        </thead>
        <tbody
          class="ons-table__body"
        >
          <tr
            class="ons-table__row"
            data-testid="questionnaire-table-row"
          >
            <td
              class="ons-table__cell"
            >
              <a
                href="/questionnaires/OPN2101A/cases"
              >
                OPN2101A
              </a>
            </td>
            <td
              class="ons-table__cell"
            >
              gusty
            </td>
            <td
              class="ons-table__cell"
            >
              0
            </td>
          </tr>
          <tr
            class="ons-table__row"
            data-testid="questionnaire-table-row"
          >
            <td
              class="ons-table__cell"
            >
              <a
                href="/questionnaires/OPN2007T/cases"
              >
                OPN2007T
              </a>
            </td>
            <td
              class="ons-table__cell"
            >
              gusty
            </td>
            <td
              class="ons-table__cell"
            >
              10
            </td>
          </tr>
          <tr
            class="ons-table__row"
            data-testid="questionnaire-table-row"
          >
            <td
              class="ons-table__cell"
            >
              <a
                href="/questionnaires/LMS2101_AA1/cases"
              >
                LMS2101_AA1
              </a>
            </td>
            <td
              class="ons-table__cell"
            >
              gusty
            </td>
            <td
              class="ons-table__cell"
            >
              0
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table
      class="ons-table"
      data-testid="questionnaire-table"
      id="questionnaire-table"
    >
      <thead
        class="ons-table__head"
      >
        <tr
          class="ons-table__row"
        >
          <th
            class="ons-table__header"
            scope="col"
          >
            Questionnaire name
          </th>
          <th
            class="ons-table__header"
            scope="col"
          >
            Server park
          </th>
          <th
            class="ons-table__header"
            scope="col"
          >
            Number of cases
          </th>
        </tr>
      </thead>
      <tbody
        class="ons-table__body"
      >
        <tr
          class="ons-table__row"
          data-testid="questionnaire-table-row"
        >
          <td
            class="ons-table__cell"
          >
            <a
              href="/questionnaires/OPN2101A/cases"
            >
              OPN2101A
            </a>
          </td>
          <td
            class="ons-table__cell"
          >
            gusty
          </td>
          <td
            class="ons-table__cell"
          >
            0
          </td>
        </tr>
        <tr
          class="ons-table__row"
          data-testid="questionnaire-table-row"
        >
          <td
            class="ons-table__cell"
          >
            <a
              href="/questionnaires/OPN2007T/cases"
            >
              OPN2007T
            </a>
          </td>
          <td
            class="ons-table__cell"
          >
            gusty
          </td>
          <td
            class="ons-table__cell"
          >
            10
          </td>
        </tr>
        <tr
          class="ons-table__row"
          data-testid="questionnaire-table-row"
        >
          <td
            class="ons-table__cell"
          >
            <a
              href="/questionnaires/LMS2101_AA1/cases"
            >
              LMS2101_AA1
            </a>
          </td>
          <td
            class="ons-table__cell"
          >
            gusty
          </td>
          <td
            class="ons-table__cell"
          >
            0
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
